fix: Remote file transfer strategy refactoring

hpohekar opened the pull request:
closes https://github.com/ansys/pyfluent/issues/2710 https://github.com/ansys/pyfluent/issues/2711 https://github.com/ansys/pyfluent/issues/2721

1. Remote FTS was running in the background even after calling `session.exit()` or `session.force_exit()` in the wsl. Now it is deleted after calling `session.exit()` or `session.force_exit()`
2. Nightly Dev Test Fix - Runners are subject to change for each CI therefore this scenario is handled now.
3. More configuration options are provided for remote file transfer startegy.

raph-luc commented on src/ansys/fluent/core/utils/file_transfer_service.py:
```diff
@@ -110,24 +110,29 @@ def download(
+            else [str(file_name)]
+        )
+    elif isinstance(file_name, list):
+        if os.path.exists(path):
```
What is going on here? Seems unrelated to the changes described in the PR. So `path` was hardcoded to work only in GitHub runners previously?

Also, `pathlib.Path` should be able to handle `str` paths with no issues in my understanding, do we need two different `if` clauses for `str` and `Path`?

Lastly, usually we don't need to convert a `pathlib.Path` to `str` as in line 126, in my understanding most Python tools should work with `pathlib.Path` out of the box (including all built-in functions like `open`).

hpohekar replied:
There was a bug in locating files in GitHub (container mode) as the path was hard-coded. We have resolved this now on the basis of Fluent mode. If the Fluent is running is container mode then it will check `host_mount_path` as `path` to locate files if that path is already exists else it will use given file name directly.

hpohekar replied:
@raph-luc 

Yes, I verified that we don't need to convert `pathlib.Path` to `str`. We can use it directly. We will remove `str` for `pathlib.Path`.  Thank you.

```
>>> file_path = pathlib.Path("D:/Repos/pyfluent/bug_solver.cas.h5")                            
>>> file_path                                                                                  
WindowsPath('D:/Repos/pyfluent/bug_solver.cas.h5')
>>> from ansys.fluent.core import launch_fluent 
>>> solver = launch_fluent()
>>> solver.file.read_case(file_name=file_path)
pyfluent.settings_api WARNING: Mismatch between generated file and server object info. Dynamically created settings classes will be used.
Fast-loading "D:\ANSYSDev\vNNN\fluent\fluent24.2\\addons\afd\lib\hdfio.bin"
Done.
Multicore SMT processors detected. Processor affinity set!

Reading from AAPQ13pUvWxQtxN:"D:\Repos\pyfluent\bug_solver.cas.h5" in NODE0 mode ...
  Reading mesh ...
       17822 cells,     1 cell zone  ...
          17822 mixed cells,  zone id: 87
       91581 faces,     7 face zones ...
           2168 polygonal wall faces,  zone id: 34
            268 polygonal wall faces,  zone id: 33
            155 polygonal pressure-outlet faces,  zone id: 32
            152 polygonal velocity-inlet faces,  zone id: 31
             55 polygonal velocity-inlet faces,  zone id: 30
           2001 polygonal symmetry faces,  zone id: 29
          86782 mixed interior faces,  zone id: 89
       66417 nodes,     1 node zone  ...
  Done.


Building...
     mesh
     materials,
     interface,
     domains,
        mixture
     zones,
        wall-elbow
        wall-inlet
        outlet
        cold-inlet
        hot-inlet
        symmetry-xyplane
        interior--elbow-fluid
        elbow-fluid
     parallel,
Done.
>>> 
```

hpohekar replied:
Done. Thank you.

raph-luc reviewed (commented):
Do we need these being printed to the user console? 
![image](https://github.com/ansys/pyfluent/assets/132297401/58d56a0f-7cce-45e5-bb04-71f48f1af2ec)

I believe these should only show up in debug logging calls

hpohekar commented:
> Do we need these being printed to the user console? ![image](https://private-user-images.githubusercontent.com/132297401/323607056-58d56a0f-7cce-45e5-bb04-71f48f1af2ec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM0NDU4NjAsIm5iZiI6MTcxMzQ0NTU2MCwicGF0aCI6Ii8xMzIyOTc0MDEvMzIzNjA3MDU2LTU4ZDU2YTBmLTdjY2UtNDVlNS1iYjA0LTcxZjQ4ZjFhZjJlYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxOFQxMzA2MDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MDU0MmRhM2E1NGU3Mzg2NjZjYjMyYWZmZTAxZmY5OGJmYjViODM3NDRiZTcxMDQxOWZmNzc1OTg3OWU1MDY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9FJpzbQ3ofmaWLpkis6yiHJ9xCM6k3cW-evqg2sy3lc)
> 
> I believe these should only show up in debug logging calls

Point 1 - related to `[2024-04-18 08:12:49.360737] [0x0000000000de8500] [info]    File transfer server started.`

We had a user requirement to get information about file upload/download progress therefore we are running this gRPC FTS container without `detach` mode.

Point 2 - related to 

```
>>> solver.exit()
>>> 67bfe7fb1fe2
```

This is docker's default behavior. Docker return ID of container when we use `docker kill <id>`

hpohekar commented on src/ansys/fluent/core/session.py:
```diff
@@ -326,11 +326,19 @@ def exit(self, **kwargs) -> None:
         """Immediately terminates the Fluent session, losing unsaved progress and
         data."""
         self._fluent_connection.force_exit()
+        try:
+            self._file_transfer_service.exit()
+        except AttributeError:
+            pass
```
Use is allowed to plug in own FTS. If user has not implemented method to exit respected FTS then we will get `Attribute` error. To handle this scenario we have done this.

raph-luc replied:
Presumably if users set up their own FTS, then they would have to provide an `exit` method as well, right? Otherwise they will constantly have zombie services running in their machine. 

Doesn't that mean that we don't want to hide an `AttributeError` here, so that the user is aware of this issue?

hpohekar replied:
Earlier I had kept it as `self._file_transfer_service.exit()` only without try block. We can raise an exception with more information in `except` block instead of `pass` only.

hpohekar replied:
they would have to provide an exit method as well, right? - yes.

hpohekar replied:
We have removed try block and user will get `AttributeError` now.

hpohekar replied:
We will update `exit` method. Thank you.

raph-luc commented:
> This is docker's default behavior. Docker return ID of container when we use `docker kill <id>`

The [Docker API](https://docker-py.readthedocs.io/en/stable/) provides a container stop method that doesn't do this, can we use that or does the current implementation not allow this?

hpohekar commented:
> > This is docker's default behavior. Docker return ID of container when we use `docker kill <id>`
> 
> The [Docker API](https://docker-py.readthedocs.io/en/stable/) provides a container stop method that doesn't do this, can we use that or does the current implementation not allow this?

@raph-luc 

I have tried `docker stop` but gRPC FTS is not terminated, it is still running. It terminates only using `docker kill`

![image](https://github.com/ansys/pyfluent/assets/106588300/c1dba38b-10f2-4c21-b0f6-0b88dea8dc26)

![image](https://github.com/ansys/pyfluent/assets/106588300/9e6bb543-87a1-4bb4-aa6a-c45d6cd686c8)

Thanks for the suggestion.

hpohekar commented:
> > > This is docker's default behavior. Docker return ID of container when we use `docker kill <id>`
> > 
> > 
> > The [Docker API](https://docker-py.readthedocs.io/en/stable/) provides a container stop method that doesn't do this, can we use that or does the current implementation not allow this?
> 
> @raph-luc
> 
> I have tried `docker stop` but gRPC FTS is not terminated, it is still running. It terminates only using `docker kill`
> 
> ![image](https://private-user-images.githubusercontent.com/106588300/323632028-c1dba38b-10f2-4c21-b0f6-0b88dea8dc26.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM0NDkxMDcsIm5iZiI6MTcxMzQ0ODgwNywicGF0aCI6Ii8xMDY1ODgzMDAvMzIzNjMyMDI4LWMxZGJhMzhiLTEwZjItNGMyMS1iMGY2LTBiODhkZWE4ZGMyNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxOFQxNDAwMDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZmQxZDYyZGFjM2FlYTQxNjM0NTA0ODQyNDY5YWMxYjg1OWY1NjViZmRlZDNlYzk3YWYzZWM3NWUwNTZjYWM2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.IsdgL79W38ANr8WQSa9ZEh4iixcCRTIErw0cCSjKHw8)
> 
> ![image](https://private-user-images.githubusercontent.com/106588300/323632577-9e6bb543-87a1-4bb4-aa6a-c45d6cd686c8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM0NDkxMDcsIm5iZiI6MTcxMzQ0ODgwNywicGF0aCI6Ii8xMDY1ODgzMDAvMzIzNjMyNTc3LTllNmJiNTQzLTg3YTEtNGJiNC1hYTZhLWM0NWQ2Y2Q2ODZjOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxOFQxNDAwMDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MjlhZThlZjE3NWRmY2YwYmQ1ZWI5N2RmNjVjODBmZjI0MjlkYWY0NDI4NmRiMTBlNGUyZDI3NDI1NmE5MjAxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.PSLBHS1DgLgZeoCEYbGOV8u5ISxu7VPTTasxV-0v4JM)
> 
> Thanks for the suggestion.

@raph-luc Ok. It is terminated now.

`docker stop` takes some time where as `docker kill` terminates it quickly. Ultimately both uses SIGKILL behind the scene. Thank you.

raph-luc commented on src/ansys/fluent/core/utils/file_transfer_service.py:
```diff
@@ -257,7 +257,7 @@ def exit(self):
         ]
         for container_id in active_container_ids:
             try:
-                subprocess.Popen(f"docker kill {container_id}", shell=True)
+                subprocess.Popen(f"docker stop {container_id}", shell=True)
             except Exception:
                 pass
```
This entire `exit` method definition can be replaced with Python Docker API calls (starting from line 247), which is going to be much cleaner and less error prone than parsing shell output like this. I think this PR may be a good opportunity to review this `exit` method (or in a separate PR when time permits)

hpohekar replied:
@raph-luc It's update now. We are using just `self.server.kill()` with help of Docker SDK. Thank you.

mkundu1 commented on src/ansys/fluent/core/session.py:
```diff
@@ -326,11 +326,19 @@ def exit(self, **kwargs) -> None:
+        try:
+            self._file_transfer_service.exit()
+        except AttributeError:
+            pass
```
Yes, this should be in the finalizer, otherwise the issue will still remain when the session object is garbage collected. @hpohekar

hpohekar replied:
Yes, this is the remaining thing only. We will implement this soon. Thank you.

hpohekar replied:
Done. Thank you.

hpohekar commented:
> Do we need these being printed to the user console? ![image](https://private-user-images.githubusercontent.com/132297401/323607056-58d56a0f-7cce-45e5-bb04-71f48f1af2ec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM0NTYzNDUsIm5iZiI6MTcxMzQ1NjA0NSwicGF0aCI6Ii8xMzIyOTc0MDEvMzIzNjA3MDU2LTU4ZDU2YTBmLTdjY2UtNDVlNS1iYjA0LTcxZjQ4ZjFhZjJlYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxOFQxNjAwNDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hOGQ4YTVhZDM0ODgzODNiNzlkN2ZjNzQ0ODQ2NzQ1ODMzNzdhNjcwODY0ZGI1NTA1NWZhNzE1NGM3NzllZDJlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bMwIepch68lGxZT0KH-M-Ty3cnNMzVA-Vn4dBcURbjI)
> 
> I believe these should only show up in debug logging calls

@raph-luc This is not appearing as we are using Docker SDK now. Thank you.

hpohekar commented:
> > This is docker's default behavior. Docker return ID of container when we use `docker kill <id>`
> 
> The [Docker API](https://docker-py.readthedocs.io/en/stable/) provides a container stop method that doesn't do this, can we use that or does the current implementation not allow this?

@raph-luc Done. Thank you.

raph-luc commented on src/ansys/fluent/core/utils/file_transfer_service.py:
```diff
@@ -139,26 +141,46 @@ class RemoteFileTransferStrategy(FileTransferStrategy):
-                f"docker run -p {self.host_port}:50000 -v {self.host_mount_path}:{self.container_mount_path} ghcr.io/ansys/tools-filetransfer:latest",
-                shell=True,
+            self.ports = ports if ports else {"50000/tcp": self.host_port}
+            self.server = self.docker_client.containers.run(
```
Minor: I think `container` would be a better variable name
```suggestion
            self.container = self.docker_client.containers.run(
```

hpohekar replied:
Done. Thank you.

raph-luc commented on src/ansys/fluent/core/utils/file_transfer_service.py:
```diff
@@ -240,20 +262,7 @@ def download(
-                subprocess.Popen(f"docker kill {container_id}", shell=True)
-            except Exception:
-                pass
+        self.server.kill()
```
I believe we should use `stop`  instead, as it already has a built-in configurable timeout of 10 seconds, after which it will send a kill signal (see [docs](https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.Container.stop)).

This together with the other change I suggested above would mean:
```suggestion
        self.container.stop()
```

hpohekar replied:
Done. Thank you.

raph-luc commented on src/ansys/fluent/core/utils/file_transfer_service.py:
```diff
@@ -110,25 +110,27 @@ def download(
-            files = [str(file_name)]
-        elif isinstance(file_name, list):
-            files = [str(file) for file in file_name]
+    if isinstance(file_name, str):
+        file_path_check = os.path.join(path, os.path.basename(file_name))
+        files = (
+            [str(file_path_check)] if os.path.isfile(file_path_check) else [file_name]
+        )
+    elif isinstance(file_name, pathlib.PurePath):
+        file_path_check = os.path.join(path, file_name.name)
+        files = [file_path_check] if os.path.isfile(file_path_check) else [file_name]
```
As mentioned earlier, I think we can still put these two `if` clauses together (`str` and  `Path`), by for example preparing the path with `file_name = Path(file_name)`

hpohekar replied:
Done. Thank you.

mkundu1 commented:
Not sure if the permission issue in self-hosted runs
```
PermissionError: [Errno 13] Permission denied: '/home/ansys/.local/share/ansys_fluent_core/examples/serverinfo-ggcxfjqg.txt'
```
started coming from this PR. Probably the permission of EXAMPLES_PATH directory is somehow changed.

PipKat commented on src/ansys/fluent/core/session.py:
```diff
@@ -326,11 +326,13 @@ def exit(self, **kwargs) -> None:
+        self._file_transfer_service.exit()
 
     def force_exit(self) -> None:
         """Immediately terminates the Fluent session, losing unsaved progress and
```
```suggestion
        """Immediately terminate the Fluent session, losing unsaved progress and
```

mkundu1 commented:
As I have tested, the file-transfer container doesn't support a graceful stop. I passed a large value to the `-t` parameter of `docker stop` and it is waiting for that many seconds. So, we can replace `docker stop` with `docker kill`, because `docker stop` will anyway do a forced stop (equivalent to `docker kill`) after a default grace period. This will reduce the time to exit() and will probably fix the failing test.

PipKat commented on src/ansys/fluent/core/utils/file_transfer_service.py:
```diff
@@ -150,37 +151,40 @@ def __init__(
-            Port for file transfer service to use.
+            Port for the file transfer service to use.
         container_mount_path: Union[str, Path], optional
             Path inside the container where host mount path will be mounted to.
```
```suggestion
            Path inside the container to mount the host mount path to.
```

raph-luc replied:
@PipKat I am not sure how, it is literally "mounting" the "host mount path" inside the container. Maybe:
```suggestion
            Path inside the container for the host mount path.
```
Leaving the fact that it will be mounted implied (should be enough I think).

hpohekar replied:
Done. Thank you.

PipKat commented on src/ansys/fluent/core/utils/file_transfer_service.py:
```diff
@@ -150,37 +151,40 @@ def __init__(
         container_mount_path: Union[str, Path], optional
             Path inside the container where host mount path will be mounted to.
         host_mount_path: Union[str, Path], optional
             Existing path in the host operating system that will be available inside the container.
```
```suggestion
            Existing path in the host operating system to be available inside the container.
```

hpohekar replied:
Done. Thank you.

PipKat commented on src/ansys/fluent/core/fluent_connection.py:
```diff
@@ -333,6 +335,8 @@ def __init__(
             PyPIM. This instance will be deleted when calling
             ``Session.exit()``.
+        file_transfer_service : optional
+            File transfer service. Uploads/downloads files to/from the server.
```
```suggestion
            File transfer service for uploading files to and
            downloading files from the server.
```
Avoid using slashes and meaning is often ambiguous.

hpohekar replied:
Done. Thank you.

PipKat commented on src/ansys/fluent/core/session.py:
```diff
@@ -328,7 +332,7 @@ def exit(self, **kwargs) -> None:
 
     def force_exit(self) -> None:
-        """Immediately terminates the Fluent session, losing unsaved progress and
+        """Immediately terminate the Fluent session, losing unsaved progress and
```
```suggestion
        """Immediately exit the Fluent session, losing unsaved progress and
```
I'd prefer the verb be exit to match the verb in the method name.

raph-luc replied:
@PipKat there is a difference in meaning here, `exit` is a "graceful" exit, while `terminate` is an "externally forced" exit. Do you think something like the suggestion below would be better?
```suggestion
        """Forces the Fluent session to exit, losing unsaved progress and
```
Thanks for the help!

hpohekar replied:
Done. Thank you.

PipKat commented on src/ansys/fluent/core/utils/file_transfer_service.py:
```diff
@@ -134,26 +134,55 @@ class RemoteFileTransferStrategy(FileTransferStrategy):
         host_mount_path: Optional[str] = None,
     ):
-        self.host_port = random.randint(5000, 6000)
+        """Provides gRPC-based remote file transfer strategy.
```
```suggestion
        """Provides the gRPC-based remote file transfer strategy.
```

hpohekar replied:
Done. Thank you.

raph-luc commented:
@hpohekar @PipKat note that some of these doc suggestions are for docstrings copied from `src\ansys\fluent\core\launcher\fluent_container.py`. The final docstring updates should therefore be made in both locations (as applicable). Perhaps we can do that in a separate PR to avoid bundling unrelated changes in this PR.

PipKat commented:
> @hpohekar @PipKat note that some of these doc suggestions are for docstrings copied from `src\ansys\fluent\core\launcher\fluent_container.py`. The final docstring updates should therefore be made in both locations (as applicable). Perhaps we can do that in a separate PR to avoid bundling unrelated changes in this PR.

I'm fine with that. I assume that the changes only need to made in `src\ansys\fluent\core\launcher\fluent_container.py` and will then somehow eventually get copied to here? Is this something that you need or want me to do--or will you or the PyFluent team take care of it?

raph-luc commented:
@PipKat we will take care of it, what I meant is the changes to the docstrings being made here should later be applied to the location where these docstrings were copied from (`fluent_container.py`).

hpohekar commented:
> As I have tested, the file-transfer container doesn't support a graceful stop. I passed a large value to the `-t` parameter of `docker stop` and it is waiting for that many seconds. So, we can replace `docker stop` with `docker kill`, because `docker stop` will anyway do a forced stop (equivalent to `docker kill`) after a default grace period. This will reduce the time to exit() and will probably fix the failing test.

Yes, right. All tests are passing after using 'docker kill`. Thanks a lot for having look into it.

hpohekar commented:
> Not sure if the permission issue in self-hosted runs
> 
> ```
> PermissionError: [Errno 13] Permission denied: '/home/ansys/.local/share/ansys_fluent_core/examples/serverinfo-ggcxfjqg.txt'
> ```
> 
> started coming from this PR. Looks like the permission of EXAMPLES_PATH directory is somehow changed:
> 
> ```
> ansys@pyfluent-ci-3:~$ ls -ltr /home/ansys/.local/share
> total 20
> drwxr-xr-x 3 ansys ansys 4096 Feb  9  2023 pyvista
> drwxr-xr-x 2 ansys ansys 4096 Jul 31  2023 ansys_tools_path
> drwxr-xr-x 3 ansys ansys 4096 Sep 22  2023 ansys_mapdl_core
> drwxr-xr-x 4 ansys ansys 4096 Mar 29 10:09 virtualenv
> drwxr-xr-x 4 ansys ansys 4096 Apr 18 19:19 ansys_fluent_core
> ansys@pyfluent-ci-3:~$ ls -ltr /home/ansys/.local/share/ansys_fluent_core/
> total 44
> drwxr-xr-x 4 ansys ansys 36864 Jun 13  2023 examples.bkp
> drwxr-xr-x 2 root  root   4096 Apr 18 19:19 examples
> ```
> 
> I have cleaned up the examples path within the runners, will keep checking what is going on.

@mkundu1 

This is happened again.

![image](https://github.com/ansys/pyfluent/assets/106588300/bf988c60-e180-47c0-a516-a80318c1bfc6)

I have cleared `examples` folder and changed it's read-write permission then this issue is resolved.

No idea why this is happening to this PR particularly.
